controllers/controller_test.go

```diff
@@ -1,5 +1,307 @@
-// +build todo
-
 package controllers
 
-import ()
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+// E2Eish test against test env
+func TestController(t *testing.T) {
+	// Setup separate test env and start controller
+	req := require.New(t)
+
+	timeout := time.Second * 10
+	duration := time.Second * 4
+	interval := time.Millisecond * 300
+
+	testEnv = &envtest.Environment{}
+	conf, err := testEnv.Start()
+	req.Nil(err)
```

Comment:
I recommend having a look at https://github.com/vshn/k8up/blob/master/controllers/envsuite_test.go
It's based on Test suites: https://pkg.go.dev/github.com/stretchr/testify/suite

This allows much cleaner and more isolated testing :)

Reply:
I honestly found the suite setup a lot more confusing. If you prefer switching to suite I can, but all other tests really just need a k8s client and operate an their own namespace, so isolation is not really a problem IMO.

Reply:
tbh I'd prefer some sort of making smaller tests. The current test in `controllers/backup_test.go ` for example is a single func with 250 lines of code, that doesn't look maintainable at all

Reply:
I agree, the tests got too large. Do you have a concrete suggestion on how to fix that? Simply switching to `testify/suite` doesn't change that.  I'm really not sure how to fix this.

Reply:
You could start incrementally. Refactor our the boilerplate code to setup and teardown the envtest environment. The Suite would be perfect use case for this.
Continue then with extracting blocks of codes into their own funcs, e.g. certain assertion helpers. At least that would make tests a bit more readable by using natural English language.
An example could be this one: https://github.com/vshn/k8up/blob/master/controllers/check_integration_test.go#L63

Going further I would try to find ways to split test cases. Sometimes a little bit of copy-paste in the test code resp. assertions is acceptable if it helps cleaning up. Also, nothing stops you from creating multiple test files that could maybe test different scenarios or even large business functions.